### ISSUE TITLE
Extend Databricks Operators to DBFS Interaction

### ISSUE THREAD

**Issue by nivangio** [apache/airflow#39262]
### Description

Create operators and Hook to interact with Databricks' DBFS (https://docs.databricks.com/api/workspace/dbfs)

### Use case/motivation

As per latest databricks plugin (https://github.com/apache/airflow/tree/main/airflow/providers/databricks) there is no possibility to interact with DBFS API.

As I had to do it in my job (and I have it quite developed), I thought it'd be a good idea to share it with the community

So far, I've got:

- An operator that uploads files to DBFS
- A hook that interacts with the DBFS API, respecting Databricks' Hooks logic and inheriting from `BaseDatabricksHook`

As part of the PR, I'd add:

- Some more operators (getting files, getting files metadata, deleting files)
- Tests in line with Airflow's test suite

Please LMK if you consider this a relevant contribution or not

### Related issues

As one of the DBFS API endpoints uses `PUT` as verb., I'd need to include a modification in `BaseDatabricksHook`, because it is not supporting PUT ATM (see https://github.com/apache/airflow/blob/main/airflow/providers/databricks/hooks/databricks_base.py#L584)

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by Taragolis** [comment]
Maybe it is also good idea to implements DBFS over [Object Storage](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/objectstorage.html)

**Comment by Taragolis** [comment]
And just wondering why not implement it over the [official SDK](https://databricks-sdk-py.readthedocs.io/en/latest/workspace/files/dbfs.html)?

Note about production usage from https://docs.databricks.com/en/dev-tools/sdk-python.html

> [!NOTE] 
> This feature is in [Beta](https://docs.databricks.com/en/release-notes/release-types.html) and is okay to use in production.
>
>During the Beta period, Databricks recommends that you pin a dependency on the specific minor version of the Databricks SDK for Python that your code depends on.

**Comment by nivangio** [comment]
Hi @Taragolis . Thx for your reply! TBH, I wasn't aware of the existence of Object Storage. It seems as if many of the things I've implemented were already there. The only thing I cannot find is some sort of `cp` that enables uploading/downloading data from DBFS. At this point I wonder if it wouldn't be better to extend [this](https://filesystem-spec.readthedocs.io/en/stable/_modules/fsspec/implementations/dbfs.html) and then use it within `ObjectStoragePath`. 

With respect to the SDK, sounds good to me. However, the whole plugin is done pointing directly to the REST Endpoints. I think it may be better in that sense to stick to one strategy (either change everything to point to the SDK or extend it using the REST API)

**Comment by Taragolis** [comment]
> if it wouldn't be better to extend [this](https://filesystem-spec.readthedocs.io/en/stable/_modules/fsspec/implementations/dbfs.html) and then use it within ObjectStoragePath

AIrflow ObjectStorage build in top of the `fsspec` and I guess extend some methods, like [copy](https://github.com/apache/airflow/blob/191b5c30e68566a75f67aefc860f59573b79bed6/airflow/io/path.py#L261)

> With respect to the SDK, sounds good to me. However, the whole plugin is done pointing directly to the REST Endpoints

Small nit, this one about [Airflow Provider](https://airflow.apache.org/docs/apache-airflow-providers/), not a [Airflow Plugin](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/plugins.html#plugins) that is a bit different things.

> I think it may be better in that sense to stick to one strategy (either change everything to point to the SDK or extend it using the REST API) 

In the long run SDK should replace internal solutions, that is why I propose to use SDK over the direct call to the API

**Comment by nivangio** [comment]
> In the long run SDK should replace internal solutions, that is why I propose to use SDK over the direct call to the API

Absolutely agree on the idea! I think that's a quite deep change and I am not sure how that's handled  and if it shouldn't be actually part of another ticket (i.e., more of a refactor ticket than a feature add one)

**Comment by nivangio** [comment]
@Taragolis @eladkal should I move forward with this as originally posted or do you have sth different in mind?

**Comment by moomindani** [comment]
This has been idle since May 2024, with the original question about scope never answered. Two things suggest closing rather than reviving it.

**What this issue targets is deprecated.** Per [the DBFS docs](https://docs.databricks.com/aws/en/dbfs/), "Both DBFS root and DBFS mounts are deprecated and not recommended by Databricks" — and those are exactly what the [DBFS API](https://docs.databricks.com/api/workspace/dbfs) this issue proposes wrapping is used to read and write. Building a new operator surface on top of that would point users at a path Databricks steers away from. Worth being precise, though: this is not a blanket deprecation of DBFS as a concept — the `dbfs:` URI scheme remains central, and disabling root/mounts does not disable the URI itself. It is the storage locations this API operates on that are deprecated, not the identifier.

**There is no demand signal.** This is the only DBFS issue in the repo, and I could not find any request for Unity Catalog volumes file operations either — no issues, no PRs. So nothing indicates users are blocked on either API.

If someone does need volumes file operations later, Airflow's `ObjectStoragePath` / fsspec integration (declared via `filesystems:` in `provider.yaml`, as amazon, google, and sftp do) may fit better than a bespoke operator set, since volumes expose a `/Volumes/<catalog>/<schema>/<volume>/...` path model. That would be a fresh issue with a concrete use case behind it, not this one.

Suggest closing as stale/superseded.

---
Drafted-by: Claude Code (Opus 5); reviewed by @moomindani before posting
